### ISSUE TITLE
you should update your version of rollinghash

### ISSUE THREAD

**Issue by chmduquesne** [syncthing/syncthing#4693]
Hi there,

Author of https://github.com/chmduquesne/rollinghash here.

I know that syncthing uses rollinghash for some operations, and I see that vendoring is used.
- The version you use is quite old. I added some optimization tricks in recent versions and it should be quite faster to roll on a file. The new version lets the compiler take advantage of the returned type and inline the Roll() function. On my own benchmarks, I observed up to x4 speed.
- I see that you use adler32. I recommend giving buzhash32 a try instead, I expect it to be marginally faster and to have less collisions.

Cheers!

**Comment by imsodin** [comment]
We are on the second newest commit since https://github.com/syncthing/syncthing/commit/c24bf7ea55c1bc75043fe50a5adec4da9b9941da / v0.14.43 and the newest one only concerns docs, so we should already be profiting from your optimizations - thanks a lot for those.

I ran buzhash32 through our existing benchmark for adler32 and it looks promising:
```
BenchmarkWeakHashAdler32Roll-4              2000            819669 ns/op        159.91 MB/s
BenchmarkWeakHashbuzhash32Roll-4            2000            569554 ns/op        230.13 MB/s
```

@AudriusButkevicius Was there a reason not to use that or simply not available at the time?
I think changing would need some kind of transition, but if the above is representative for a wide range of devices that might be worth it.

**Comment by calmh** [comment]
If we wanted to change I think we could just do so. What’ll happen is we won’t see matches across the transition, which is fine, or get false positives which are caught by the real hash and are also fine.

On 20 Jan 2018, at 19:54, Simon Frei <notifications@github.com<mailto:notifications@github.com>> wrote:


We are on the second newest commit since c24bf7e<https://github.com/syncthing/syncthing/commit/c24bf7ea55c1bc75043fe50a5adec4da9b9941da> / v0.14.43 and the newest one only concerns docs, so we should already be profiting from your optimizations - thanks a lot for those.

I ran buzhash32 through our existing benchmark for adler32 and it looks promising:

BenchmarkWeakHashAdler32Roll-4              2000            819669 ns/op        159.91 MB/s
BenchmarkWeakHashbuzhash32Roll-4            2000            569554 ns/op        230.13 MB/s


@AudriusButkevicius<https://github.com/audriusbutkevicius> Was there a reason not to use that or simply not available at the time?
I think changing would need some kind of transition, but if the above is representative for a wide range of devices that might be worth it.

**Comment by chmduquesne** [comment]
Ah ok, if it was done 22 days ago it should be fine, these optimizations were done slightly before christmas :-)

**Comment by AudriusButkevicius** [comment]
I agree we can just switch over and have misses during the transition period.

**Comment by AudriusButkevicius** [comment]
Interestingly I get:

```
BenchmarkWeakHashAdler32-4           	   20000	     81974 ns/op	1598.93 MB/s
BenchmarkWeakHashAdler32Roll-4       	    2000	    871249 ns/op	 150.44 MB/s
BenchmarkWeakHashRabinKarp64-4       	       1	2570001500 ns/op	   0.05 MB/s
BenchmarkWeakHashRabinKarp64Roll-4   	    2000	    808250 ns/op	 162.17 MB/s
BenchmarkWeakHashBozo32-4            	    3000	    452666 ns/op	 289.56 MB/s
BenchmarkWeakHashBozo32Roll-4        	    3000	    496999 ns/op	 263.73 MB/s
BenchmarkWeakHashBuzhash32-4         	    5000	    321999 ns/op	 407.06 MB/s
BenchmarkWeakHashBuzhash32Roll-4     	    3000	    597667 ns/op	 219.31 MB/s
BenchmarkWeakHashBuzhash64-4         	    5000	    324800 ns/op	 403.55 MB/s
BenchmarkWeakHashBuzhash64Roll-4     	    3000	    583167 ns/op	 224.76 MB/s
```

I guess the most relevant part is the rolling part, yet in my case bozo wins.

**Comment by imsodin** [comment]
When recommending buzhash @chmduquesne mentioned both speed and collisions - maybe bozo is significantly worse regarding collisions?

**Comment by chmduquesne** [comment]
Bozo should be ok in terms of collisions too. I wrote a benchmark to check
how frequently the first n bits are equals to 1 when rolling on a random 64
bytes window, and it follows the expected pattern of a well spread checksum.

I initially wrote this code as an implementation of rabinkarp, but I
realized later that I had misunderstood the concept after reading the
original paper. However I kept it in the library because it is performant
and experimentally it has all the good properties.

In terms of maths, it follows the pattern of a congruent generator, which
is actually used in some implementations of the random() function. As a
random function, it is not so awesome because of its predictability
(although it has been successfully modified to generate the family pgc
random so it a sound base), but as a checksum I believe it should be pretty
good, since we don't care about that aspect.

Don't hesitate to use it!

**Comment by AudriusButkevicius** [comment]
@calmh just thinking out loud. Does it make sense to use some custom hash protocol wise?
Should we use buzzhash instead, so that hopefully others don't need to reimplement it, like luhn?
